<h1 align="center"><b>CORPORACIÓN UNIVERSITARIA ADVENTISTA</b></h1>
<h1 align="center"><b>FACULTAD DE INGENIERÍA</b></h1>
<h1 align="center"><b>ESPECIALIZACIÓN EN ANALÍTICA DE DATOS Y BIG DATA</b></h1>
<h2 align="center"><b>CURSO:</b></h1>
<h2 align="center"><b>LENGUAJES DE PROGRAMACIÓN PARA INTELIGENCIA DE NEGOCIOS</b></h1>
<h2 align="center"><b>Tema 2 – Manipulación y Limpieza de Datos</b></h1>
<h3 align="center">2025</h1>
<h3 align="center">MEDELLÍN - COLOMBIA </h1>

*** 
|[![Outlook](https://img.shields.io/badge/Gmail-D14836??style=plastic&logo=microsoft-outlook&logoColor=white)](mailto:docente.calvarez@unac.edu.co)||[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/carlosalvarezh/imgs/blob/main/Semana3.ipynb)
|-:|:-|--:|
|[![LinkedIn](https://img.shields.io/badge/linkedin-%230077B5.svg?style=plastic&logo=linkedin&logoColor=white)](https://www.linkedin.com/in/carlosalvarez5/)|[![@alvarezhenao](https://img.shields.io/twitter/url/https/twitter.com/alvarezhenao.svg?style=social&label=Follow%20%40alvarezhenao)](https://twitter.com/alvarezhenao)|[![@carlosalvarezh](https://img.shields.io/badge/github-%23121011.svg?style=plastic&logo=github&logoColor=white)](https://github.com/carlosalvarezh)|  
***

<table>
 <tr align=left><td><img align=left src="https://github.com/carlosalvarezh/imgs/blob/main/CCLogoColorPop1.gif?raw=true" width="55">
 <td>Text provided under a Creative Commons Attribution license, CC BY-NC 4.0</td>
 <td>All code is made available under the FSF-approved MIT license.(c) Corporación Universitaria Adventista de Colombia</td>
</table>

***

# **Semana 3 – Manipulación y Limpieza de Datos - Caso práctico**

## **Introducción**

La limpieza de datos en BI no es un trámite previo: es el proceso que permite que los indicadores de negocio sean defendibles. En entornos con múltiples orígenes (ERP, CRM, hojas de cálculo y APIs), los datos llegan con errores de captura, formatos heterogéneos y reglas de negocio aplicadas de forma desigual. El objetivo de este tema es establecer un procedimiento claro —y reproducible— para transformar esas entradas en un conjunto consistente, con trazabilidad de decisiones y apto para alimentar modelos estrella y tableros.

## **Fundamentos: calidad y confiabilidad en BI**

Trabajaremos sobre cinco propiedades observables:

- **Validez:** el dato cumple con el formato y el dominio permitido (fechas analizadas, correos con patrón razonable, códigos en catálogos oficiales).  

- **Exactitud:** el valor refleja la realidad del negocio (direcciones vigentes, moneda correcta, categoría coherente).  

- **Completitud:** están presentes los atributos requeridos para la definición del indicador o del modelo dimensional.  

- **Consistencia:** no hay contradicciones internas ni entre fuentes (la misma entidad mantiene los mismos valores bajo la misma definición).  

- **Unicidad:** cada entidad se representa una sola vez en su contexto maestro.  

Estas propiedades orientan cada decisión técnica del ciclo; no son un apéndice, sino los criterios de aceptación del dataset.


## **Ciclo de limpieza en BI**

<div align="center">
  <img src="https://github.com/carlosalvarezh/imgs/blob/main/DataCleaning.PNG?raw=true" width="350">
</div>

El ciclo que seguiremos es: importar, integrar, tratar faltantes, estandarizar, transformar (cuando aplique), desduplicar, verificar y enriquecer, exportar. Cada paso deja evidencias: qué se hizo, cuántas filas/columnas afectó y por qué.

## **Caso Práctico – Integración y Limpieza de Datos en BI**

Una empresa con operaciones nacionales está en proceso de implementar un sistema de **Business Intelligence (BI)** que permita consolidar información de distintos orígenes. Para iniciar el proyecto, se ha extraído un conjunto de archivos provenientes de tres sistemas principales:

1. **CRM – crm_clientes.csv**
   Registra la información maestra de clientes: identificadores, nombres, correos electrónicos, teléfonos, direcciones, ciudad, país, fecha de alta y segmento.
   Sin embargo, al revisar el archivo se observa la presencia de datos problemáticos:

   * Correos electrónicos mal formados o incompletos.
   * Variantes en la escritura de ciudades y países (e.g., “Medellín/Medellin”, “CO/Colombia/COL”).
   * Nombres con tildes, espacios de más o mayúsculas/minúsculas inconsistentes.
   * Registros duplicados y algunos identificadores colisionados.
   * Fechas en diferentes formatos.

2. **Soporte – soporte_altas_bajas.csv**
   Contiene el histórico de altas y bajas de clientes, generado por el sistema de soporte.
   Al inspeccionarlo se encuentran situaciones frecuentes en proyectos reales:

   * Estados escritos con diferentes variantes (“Alta”, “ALTA”, “ baja ”).
   * Motivos poco estandarizados (“Onboarding”, “N/A”, “—”).
   * Fechas en blanco o con formatos diferentes.
   * Algunos identificadores de clientes que no existen en el CRM.

3. **ERP – erp_ventas.csv**
   Incluye el detalle de las transacciones de ventas: identificador de venta, cliente, producto, fecha, cantidad, precios, impuestos, descuentos, monto y moneda.
   El archivo presenta los siguientes retos:

   * Cantidades registradas con valores no numéricos (por ejemplo, “dos”).
   * Montos expresados en distintos formatos (símbolos de moneda, punto o coma decimal, enteros como texto).
   * Producto_id faltante en ciertos registros.
   * Ventas duplicadas.
   * Monedas con variantes de escritura (“COP”, “cop ”, “USD”).

### **Objetivo del caso**

El objetivo es **integrar y limpiar estos datasets** siguiendo el ciclo propuesto en el curso: importar, integrar, tratar faltantes, estandarizar, transformar, desduplicar, verificar, enriquecer y exportar.

En esta primera fase se trabajará únicamente en la **integración básica** de las fuentes, sin preocuparse aún por duplicados, valores faltantes o inconsistencias.

### **Actividades iniciales**

0. Inspección visual de los datos para entender cómo vienen.
1. Importar los archivos en Pandas con políticas simples de lectura (tipos y valores nulos básicos).
2. Integrar los clientes con los eventos de soporte (`merge` por cliente_id).
3. Preparar la base de ventas para integrarla posteriormente con los clientes.


### **Etapa 0 - Inspección visual**

Al revisar los archivos, lo primero que llama la atención es que, aunque están en formato CSV, toda la información aparece en una única columna separada por `;`. Esto requiere un paso previo de parseo para reconstruir la estructura en columnas y facilitar el trabajo con campos separados. Identificar este aspecto desde el principio es clave, ya que impacta todo el proceso de limpieza posterior.  

También parece claro que ambos archivos contienen exactamente el mismo número de registros, lo que sugiere que no son fuentes independientes, sino complementos de un mismo conjunto de transacciones. La tabla de ventas proporciona los montos y productos, mientras que la tabla de descuentos ofrece método de pago, ubicación y la bandera de descuento. Esto implica no solo integrarlas usando la transacción como clave, sino también verificar que cada venta tenga su correspondiente, ya que un registro huérfano podría afectar la coherencia de los datos.  

Finalmente, ya desde esta primera exploración aparecen retos claros: fechas almacenadas como texto, descuentos representados como “`VERDADERO/FALSO`” o en blanco, y posibles inconsistencias entre precio, cantidad y total gastado. Estas señales muestran que el pipeline de limpieza deberá ir más allá de la corrección sintáctica, incorporando reglas de negocio que aseguren coherencia y fiabilidad en los datos finales.

#### **Parseo**

El parseo es el proceso de tomar un texto en bruto y descomponerlo siguiendo ciertas reglas conocidas para convertirlo en una estructura comprensible y manipulable por el sistema. En este caso particular, aunque los archivos se presentan como CSV, el contenido viene todo junto en una sola columna con los campos separados por punto y coma (;). Eso significa que, al cargarlos sin parámetros adicionales, Pandas lo interpreta como una única cadena larga en cada fila, perdiendo la noción de qué es el identificador de transacción, qué es el cliente, cuál es el precio unitario, etc.  

Aplicar parseo aquí consiste en indicarle explícitamente al lector de archivos cuál es el delimitador real (sep=";"). Con esa instrucción, el sistema “lee” la cadena de cada fila, reconoce dónde hay un corte de columna y asigna correctamente los valores a cada campo. De esta manera, una línea como: `TXN_6867343;CUST_09;Patisserie;Item_10_PAT;185;8;1480` se transforma en un registro con columnas diferenciadas: `Transaction ID = TXN_6867343, Customer ID = CUST_09, Category = Patisserie, Item = Item_10_PAT, Price Per Unit = 185, Quantity = 8, Total Spent = 1480`.  

En la práctica, el parseo transforma un bloque de texto plano en una tabla con tipos de datos reconocibles (como fechas, números y cadenas), sobre la cual se pueden realizar validaciones, cálculos y transformaciones. Sin este proceso, cualquier limpieza posterior sería limitada, ya que se trabajaría con un texto sin estructura en lugar de datos organizados.


### **Etapa 1 - Importación:**

El primer paso no es solo cargar archivos, sino reconocer que no están en formato estructurado. Ambos requieren **parseo por delimitador `;`** y la definición de tipos de datos adecuados:

In [32]:
import pandas as pd # importar pandas para la manipulación de datos

ventas = pd.read_csv( # cargar el archivo CSV de ventas
    "datasets/retail_store_sales0.csv", # ruta del archivo
    sep=";",  # especificar que el separador de campos es punto y coma
    dtype={ # definir tipos de datos para ciertas columnas
        "Price": "float64", # precio como número decimal
        "Transaction ID": "string", # ID de la transacción como cadena
        "Customer ID": "string", # ID del cliente como cadena
        "Category": "string", # categoría del producto como cadena
        "Item": "string" # nombre del ítem como cadena
    },
    parse_dates=False, # las fechas vienen en el otro dataset
    na_values=["", "NA", "N/A", "null"] # definir valores que deben interpretarse como nulos
)

descuentos = pd.read_csv(
    "datasets/retail_store_discount.csv",
    sep=";", 
    dtype={
        "Transaction ID": "string",
        "Payment Method": "string",
        "Location": "string"
    },
    na_values=["", "NA", "N/A", "null"]
)


En resumen: este código **lee el archivo de ventas, separa los campos correctamente, fuerza a que las claves y etiquetas se traten como texto, y marca los nulos de manera explícita**. Es el paso que convierte el archivo en bruto en un DataFrame con columnas estructuradas y tipificadas de forma adecuada para análisis y limpieza.

1. **`pd.read_csv("retail_store_sales0.csv", …)`**
   Indica a Pandas que debe leer el archivo CSV con las ventas.

2. **`sep=";"`**
   Define que el delimitador entre columnas no es la coma por defecto, sino el punto y coma (`;`). Esto permite separar correctamente cada campo en su columna.

3. **`dtype={...}`**
   Obliga a que ciertas columnas se lean como texto (`string`) aunque contengan números o identificadores alfanuméricos:

   * `Transaction ID` → se maneja como cadena para no perder ceros o sufijos.
   * `Customer ID` → lo mismo, es un código, no un número.
   * `Category` y `Item` → son etiquetas descriptivas, no deben convertirse en categorías automáticas o números.

4. **`parse_dates=False`**
   Aquí se indica explícitamente que no convierta ningún campo en fechas. La razón es que en este dataset las fechas no están, sino en el otro archivo de descuentos. Dejarlo en `False` evita errores de interpretación.

5. **`na_values=["", "NA", "N/A", "null"]`**
   Declara qué cadenas deben considerarse como valores faltantes (`NaN`). Así, si en alguna celda aparece un texto vacío, `NA`, `N/A` o `null`, Pandas lo transformará automáticamente en un valor nulo, listo para tratamiento posterior.


Una vez cargados los datos en las variables `ventas` y `descuentos`, conviene realizar una primera exploración sobre una de ellas —en este caso `ventas`— para identificar la estructura y el contenido obtenido. Como punto de partida, se pueden visualizar las primeras diez y las últimas diez filas del dataset, lo que ofrece un panorama inicial de cómo están organizados los registros.

In [33]:
ventas.head(10)

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT           185.0   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK           290.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT           215.0   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV           275.0   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD           125.0   
5    TXN_7482416     CUST_09     Patisserie          <NA>             NaN   
6    TXN_3652209     CUST_07           Food   Item_1_FOOD            50.0   
7    TXN_1372952     CUST_21      Furniture          <NA>           335.0   
8    TXN_9728486     CUST_23      Furniture   Item_16_FUR           275.0   
9    TXN_2722661     CUST_25       Butchers   Item_22_BUT           365.0   

   Quantity  Total Spent  
0     100.0       1850.0  
1      90.0       2610.0  
2      20.0        430.0  
3      90.0       2475.0  
4      70.0        875.0  
5     100.0       2000.0  
6      80.0        400.0  
7       NaN          NaN  
8      10.0        275.0  
9      30.0       1095.0

In [34]:
ventas.tail(10) 

Transaction ID Customer ID                            Category  \
12565    TXN_1712895     CUST_07  Computers and electric accessories   
12566    TXN_6298574     CUST_01                       Milk Products   
12567    TXN_4890318     CUST_25                            Butchers   
12568    TXN_1096134     CUST_06                                Food   
12569    TXN_2318174     CUST_18                          Patisserie   
12570    TXN_9347481     CUST_18                          Patisserie   
12571    TXN_4009414     CUST_03                           Beverages   
12572    TXN_5306010     CUST_11                            Butchers   
12573    TXN_5167298     CUST_04                           Furniture   
12574    TXN_2407494     CUST_23                                Food   

              Item  Price Per Unit  Quantity  Total Spent  
12565   Item_2_CEA            65.0      50.0        325.0  
12566  Item_1_MILK            50.0      40.0        200.0  
12567   Item_3_BUT            80.0      50.0        400.0  
12568  Item_4_FOOD            95.0      80.0        760.0  
12569   Item_1_PAT            50.0      90.0        450.0  
12570  Item_23_PAT           380.0      40.0       1520.0  
12571   Item_2_BEV            65.0      90.0        585.0  
12572   Item_7_BUT           140.0     100.0       1400.0  
12573   Item_7_FUR           140.0      60.0        840.0  
12574  Item_9_FOOD           170.0      30.0        510.0

El volumen de datos es tan amplio que una inspección visual simple no basta para comprenderlos en su totalidad. Por ello, utilizaremos Pandas para que nos proporcione un resumen con la información más relevante del DataFrame.

In [35]:
ventas.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Transaction ID  12575 non-null  string 
 1   Customer ID     12575 non-null  string 
 2   Category        12575 non-null  string 
 3   Item            11362 non-null  string 
 4   Price Per Unit  11966 non-null  float64
 5   Quantity        11971 non-null  float64
 6   Total Spent     11971 non-null  float64
dtypes: float64(3), string(4)
memory usage: 687.8 KB


El resultado de `ventas.info()` proporciona una visión rápida deldataset de ventas. Es un `DataFrame` con **12.575** registros y **7** columnas, en el que Pandas ya reconoció el tipo de dato de cada campo y cuántos valores no nulos hay.

- **Transaction ID** y **Customer ID** aparecen completos, con 12.575 valores no nulos, y se almacenan como cadenas de texto (`string`). Esto es coherente, ya que son identificadores alfanuméricos que no deberían tratarse como números.  

- **Category** también está completa y en formato de cadena, lo que garantiza que todas las transacciones tienen un grupo de producto asignado.  

- **Item** tiene solo 11.362 valores no nulos, lo que indica que en más de 1.200 registros el producto específico no fue registrado o se perdió en la fuente de datos.  

- **Price Per Unit**, **Quantity** y **Total Spent** tienen entre 11.966 y 11.971 valores no nulos, lo que significa que hay aproximadamente 600 transacciones sin información de precio, cantidad o monto total. Desde el punto de vista del negocio, estas filas representan un problema ya que impiden determinar el valor real de la venta.  

En cuanto a los tipos de datos, las variables monetarias y de cantidad están como float64, lo que facilita operaciones aritméticas. El consumo de memoria es de unos 688 KB, lo que es liviano y manejable.  

En resumen, la exploración preliminar indica que el conjunto de datos está bien organizado en cuanto a los tipos de variables, pero **carece de datos esenciales en los campos de producto y monto de venta**. Estas lagunas requerirán decisiones de limpieza: eliminar transacciones incompletas, imputar valores según reglas de negocio o marcar las entradas para revisión.

A continuación podemos emplear el método `.describe()` para que muestre los estadísticos básicos de las variables numéricas presentes en el dataset. Esto puede proporcionar una idea de patrones, escalas o posibles anomalías en los datos. 

In [36]:
ventas.describe()

Price Per Unit      Quantity   Total Spent
count    11966.000000  11971.000000  11971.000000
mean       233.659117     55.363796   1296.525771
std        107.435190     28.578828    947.506967
min         50.000000     10.000000     50.000000
25%        140.000000     30.000000    510.000000
50%        230.000000     60.000000   1085.000000
75%        335.000000     80.000000   1920.000000
max        410.000000    100.000000   4100.000000

En este caso, sobre el archivo de `ventas`, se observa que el **precio unitario** oscila entre 50 y 410, con una media cercana a 234 y una dispersión moderada (desviación estándar ≈ 107). La **cantidad** de productos vendidos por transacción varía de 10 a 100, con un promedio de 55 y un rango intercuartílico relativamente amplio, lo que indica heterogeneidad en los volúmenes de compra. Por su parte, el **gasto total** por transacción se sitúa en promedio en 1.296, aunque puede llegar hasta 4.100, lo cual refleja operaciones de gran magnitud. Este análisis preliminar no solo ayuda a caracterizar el comportamiento de las ventas, sino que también alerta sobre la necesidad de verificar la coherencia entre precio, cantidad y monto total, ya que estos tres campos deberían guardar una relación matemática directa.

Ahora revisemos el dataset `descuentos`:


In [37]:
descuentos.head()

Transaction ID  Payment Method Location Transaction Date Discount Applied
0    TXN_6867343  Digital Wallet   Online       08/04/2024        VERDADERO
1    TXN_3731986  Digital Wallet   Online       23/07/2023        VERDADERO
2    TXN_9303719     Credit Card   Online       05/10/2022            FALSO
3    TXN_9458126     Credit Card   Online       07/05/2022              NaN
4    TXN_4575373  Digital Wallet   Online       02/10/2022            FALSO

In [38]:
descuentos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    12575 non-null  string
 1   Payment Method    12575 non-null  string
 2   Location          12575 non-null  string
 3   Transaction Date  12575 non-null  object
 4   Discount Applied  8376 non-null   object
dtypes: object(2), string(3)
memory usage: 491.3+ KB


El comando `descuentos.info()` confirma que la tabla tiene **12.575** registros y **5** columnas. `Transaction ID`, `Payment Method` y `Location` están completos y en formato `string`, mientras que `Transaction Date` aparece como `object`, por lo que debe convertirse a `datetime` para análisis temporal. El mayor problema está en `Discount Applied`, que solo tiene **8.376** valores no nulos, dejando más de 4.000 transacciones sin información de descuento. Esto obliga a definir una estrategia de tratamiento claro antes de integrar el dataset con las ventas.

### **Etapa 2 - Integración:**

Unir los dos datasets de transacciones usando `Transaction ID` como clave, garantizando una **cardinalidad 1:1** y que no queden registros huérfanos. Nos basaremos en los siguientes **Criterios de aceptación:** (i) `Transaction ID` debe ser único en cada tabla; (ii) la unión `outer` de diagnóstico no debe tener `left_only` ni `right_only`; (iii) tras la unión final de `inner`/`left`, no debe haber duplicados ni pérdida de registros sin justificación.

**Riesgos comunes.** Espacios en las claves, tipos de datos inconsistentes, duplicados por transacción en alguna tabla y colisiones en nombres de columnas (que se solucionan con `suffixes`).

**Checklist previo**
* Clave como `string` en ambas tablas y sin espacios ocultos.  
* Unicidad de `Transaction ID` por tabla.  
* Definir sufijos si hay nombres repetidos (no parece el caso aquí, pero dejamos el patrón).  

Inicialmente, definiremos algunos pasos intermedios:


In [39]:
# Saneamiento mínimo de la clave
# asegurar string + trimming de espacios
ventas["Transaction ID"] = ventas["Transaction ID"].astype("string").str.strip()
descuentos["Transaction ID"] = descuentos["Transaction ID"].astype("string").str.strip()


In [40]:
### Diagnóstico de unicidad (por tabla)
dup_ventas = ventas["Transaction ID"].duplicated().sum()
dup_desc   = descuentos["Transaction ID"].duplicated().sum()
print({"duplicados_ventas": int(dup_ventas), "duplicados_descuentos": int(dup_desc)})


{'duplicados_ventas': 0, 'duplicados_descuentos': 0}


Ambos en 0. Esto indica que no hay registros duplicados.

Antes de hacer una unión definitiva de los datasets, haremos un paso intermedio a manera de diagnóstico para identificar posibles colisiones.

In [41]:
# Merge de diagnóstico (no definitivo)

diag = ventas.merge(
    descuentos,
    on="Transaction ID",
    how="outer",
    validate="one_to_one",
    indicator=True,
    suffixes=("", "_desc")  # por si luego hay colisiones
)
print(diag["_merge"].value_counts(dropna=False))

_merge
both          12575
left_only         0
right_only        0
Name: count, dtype: int64


El resultado del merge de diagnóstico muestra que las **12.575 filas** se clasificaron en la categoría `both`. Esto significa que **cada transacción presente en la tabla de ventas también tiene su correspondencia en la tabla de descuentos** y viceversa. En otras palabras, no existen registros huérfanos (`left_only` o `right_only` en cero) y la cardinalidad entre los dos datasets es efectivamente **uno a uno**.

Muchas veces este no será el caso, por lo que hay qué identificar cuáles son los registros "huerfanos", es decir, lo que están en una o en otra tabla, pero no en ambas. Una posible solución puede ser:

In [42]:
### Muestra de huérfanos (solo si los hay)

diag_left  = diag.loc[diag["_merge"]=="left_only", ["Transaction ID"]].head(10)
diag_right = diag.loc[diag["_merge"]=="right_only", ["Transaction ID"]].head(10)
diag_left, diag_right

(Empty DataFrame
 Columns: [Transaction ID]
 Index: [],
 Empty DataFrame
 Columns: [Transaction ID]
 Index: [])

Como habíamos adelantando, este no es el caso, por lo que el arreglo de los índices (ubicación de los datos huérfanos) está vacío, `[]`.

En términos prácticos, la integración es consistente: podemos proceder con la unión definitiva sin riesgo de duplicar ni perder registros.

In [43]:
# Unión definitiva (una vez validados los diagnósticos)

df = ventas.merge(
    descuentos,
    on="Transaction ID",
    how="inner",            # o "left" si quieres preservar todas las ventas
    validate="one_to_one",
    suffixes=("", "_desc")
)
print(df.shape)
df

(12575, 11)


Transaction ID Customer ID       Category          Item  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT           185.0   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK           290.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT           215.0   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV           275.0   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD           125.0   
...              ...         ...            ...           ...             ...   
12570    TXN_9347481     CUST_18     Patisserie   Item_23_PAT           380.0   
12571    TXN_4009414     CUST_03      Beverages    Item_2_BEV            65.0   
12572    TXN_5306010     CUST_11       Butchers    Item_7_BUT           140.0   
12573    TXN_5167298     CUST_04      Furniture    Item_7_FUR           140.0   
12574    TXN_2407494     CUST_23           Food   Item_9_FOOD           170.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0         100.0       1850.0  Digital Wallet    Online       08/04/2024   
1          90.0       2610.0  Digital Wallet    Online       23/07/2023   
2          20.0        430.0     Credit Card    Online       05/10/2022   
3          90.0       2475.0     Credit Card    Online       07/05/2022   
4          70.0        875.0  Digital Wallet    Online       02/10/2022   
...         ...          ...             ...       ...              ...   
12570      40.0       1520.0     Credit Card  In-store       03/09/2023   
12571      90.0        585.0            Cash    Online       12/08/2022   
12572     100.0       1400.0            Cash    Online       24/08/2024   
12573      60.0        840.0            Cash    Online       30/12/2023   
12574      30.0        510.0            Cash    Online       06/08/2022   

      Discount Applied  
0            VERDADERO  
1            VERDADERO  
2                FALSO  
3                  NaN  
4                FALSO  
...                ...  
12570              NaN  
12571            FALSO  
12572              NaN  
12573        VERDADERO  
12574              NaN  

[12575 rows x 11 columns]

Ya tenemos todos los datasets en un único DataFrame, df. A simple vista parece ser que no hubo problema, pero siempre es bueno realizar una verificación a posteriori.

In [44]:
# Verificaciones pos-merge

# 1) Clave sigue siendo única
print("dup_merged:", df["Transaction ID"].duplicated().sum())

# 2) Columnas esperadas y nulos introducidos
nulls_post = df.isna().sum().sort_values(ascending=False).head(10)
nulls_post

dup_merged: 0


Discount Applied    4199
Item                1213
Price Per Unit       609
Total Spent          604
Quantity             604
Transaction ID         0
Customer ID            0
Category               0
Payment Method         0
Location               0
dtype: int64

La verificación pos-merge confirma que **no hay duplicados** en `Transaction ID`. Los principales problemas son los **nulos**: `Discount Applied` (4.199), `Item` (1.213) y en los campos numéricos (`Price Per Unit`, `Quantity`, `Total Spent`) unos 600 cada uno. Los identificadores y datos de contexto están completos. Esto marca la prioridad para la siguiente etapa: definir cómo tratar esos faltantes.

### **Etapa 3 - Tratamiento de faltantes**

El propósito de esta etapa es el de gestionar los valores nulos de forma coherente con el negocio. No todos los faltantes se tratan igual: algunos son **críticos** y obligan a eliminar el registro; otros pueden **imputarse** con reglas; y otros se pueden dejar como “no aplica”, siempre que quede documentado. 

#### **Criterios de negocio**

* **Item (1.213 nulos):** si no hay producto, la transacción queda incompleta. Lo razonable es eliminar esas filas o moverlas a un dataset de excepciones.  

* **Price Per Unit, Quantity y Total Spent (~600 nulos cada uno):** son valores numéricos básicos. Si falta alguno, la venta no es válida porque no se puede calcular el monto. Estos registros deben descartarse o reconstruirse si hay coherencia entre los otros dos (ej. imputar `Total Spent = Price * Quantity`).  

* **Discount Applied (4.199 nulos):** aquí la decisión es distinta: un vacío puede interpretarse como “*no se aplicó descuento*”. La política más común es imputar con `0` (falso).

In [45]:
# Inspección rápida de nulos

df.isna().sum()

Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

Los resultados confirman que los **nulos están concentrados en las variables clave de la venta**: `Item` (1.213), `Price Per Unit` (609), `Quantity` (604) y `Total Spent` (604). El resto de campos —identificadores, categorías, método de pago, ubicación y fecha— están completos. Esto implica que la limpieza debe enfocarse en resolver la **incompletitud transaccional**, ya que sin producto, precio, cantidad o monto la venta pierde validez para análisis.


In [46]:
# Descarte de registros sin producto o sin valores de monto

df = df.dropna(subset=["Item", "Price Per Unit", "Quantity", "Total Spent"])
df.isna().sum()


Transaction ID         0
Customer ID            0
Category               0
Item                   0
Price Per Unit         0
Quantity               0
Total Spent            0
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    3783
dtype: int64

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11362 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    11362 non-null  string 
 1   Customer ID       11362 non-null  string 
 2   Category          11362 non-null  string 
 3   Item              11362 non-null  string 
 4   Price Per Unit    11362 non-null  float64
 5   Quantity          11362 non-null  float64
 6   Total Spent       11362 non-null  float64
 7   Payment Method    11362 non-null  string 
 8   Location          11362 non-null  string 
 9   Transaction Date  11362 non-null  object 
 10  Discount Applied  7579 non-null   object 
dtypes: float64(3), object(2), string(6)
memory usage: 1.0+ MB


Nos quedamos solo con transacciones que tienen todos los campos básicos completos.

In [48]:
# Imputación de descuento faltante

df["Discount Applied"] = df["Discount Applied"].fillna("FALSO")
df.isna().sum()

C:\Users\USER\AppData\Local\Temp\ipykernel_22176\2952246698.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Discount Applied"] = df["Discount Applied"].fillna("FALSO")


Transaction ID      0
Customer ID         0
Category            0
Item                0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
dtype: int64

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11362 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    11362 non-null  string 
 1   Customer ID       11362 non-null  string 
 2   Category          11362 non-null  string 
 3   Item              11362 non-null  string 
 4   Price Per Unit    11362 non-null  float64
 5   Quantity          11362 non-null  float64
 6   Total Spent       11362 non-null  float64
 7   Payment Method    11362 non-null  string 
 8   Location          11362 non-null  string 
 9   Transaction Date  11362 non-null  object 
 10  Discount Applied  11362 non-null  object 
dtypes: float64(3), object(2), string(6)
memory usage: 1.0+ MB


Actualmente, `Discount Applied` aparece como **tipo `object`**, lo que significa que sus valores se almacenan como texto heterogéneo (`VERDADERO`, `FALSO`, vacíos, incluso nulos). Trabajar con texto en un campo que en realidad representa una condición booleana dificulta cálculos y análisis: no puedes, por ejemplo, obtener tasas de descuento aplicadas o hacer agregaciones sin antes normalizar los valores.

La transformación con el `map(lambda…)` resuelve este problema al convertir la columna en una **variable binaria (0/1)**: 1 cuando hay descuento, 0 en caso contrario. De esta forma, el campo pasa de ser un texto poco estructurado a un dato cuantitativo y estandarizado, apto para análisis estadístico, validación de reglas de negocio y futuros modelos de machine learning.

Para resolver el problema de que la columna **`Discount Applied`** está en formato texto, es necesario aplicar un procedimiento que recorra cada valor y lo transforme de una representación booleana (“VERDADERO”/“FALSO”) a un valor numérico (1/0).

En este caso se utiliza **`.map()`**, un método de Pandas que permite aplicar una función sobre cada elemento de una serie o columna. Es decir, `.map()` va fila por fila y reemplaza el contenido original con el resultado de la función que se defina.

La función que se aplica es una **función `lambda`**, es decir, una función anónima y de una sola línea que no necesita nombre ni definición previa. En este caso, la lambda evalúa: si el valor convertido a texto en mayúsculas es `"VERDADERO"`, devuelve **1**; en cualquier otro caso devuelve **0**.


In [50]:
df["Discount Applied"] = df["Discount Applied"].map(
    lambda x: 1 if str(x).upper()=="VERDADERO" else 0
)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11362 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    11362 non-null  string 
 1   Customer ID       11362 non-null  string 
 2   Category          11362 non-null  string 
 3   Item              11362 non-null  string 
 4   Price Per Unit    11362 non-null  float64
 5   Quantity          11362 non-null  float64
 6   Total Spent       11362 non-null  float64
 7   Payment Method    11362 non-null  string 
 8   Location          11362 non-null  string 
 9   Transaction Date  11362 non-null  object 
 10  Discount Applied  11362 non-null  int64  
dtypes: float64(3), int64(1), object(1), string(6)
memory usage: 1.0+ MB


C:\Users\USER\AppData\Local\Temp\ipykernel_22176\831357793.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Discount Applied"] = df["Discount Applied"].map(


In [51]:
df

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT           185.0   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK           290.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT           215.0   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV           275.0   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD           125.0   
...              ...         ...            ...           ...             ...   
12570    TXN_9347481     CUST_18     Patisserie   Item_23_PAT           380.0   
12571    TXN_4009414     CUST_03      Beverages    Item_2_BEV            65.0   
12572    TXN_5306010     CUST_11       Butchers    Item_7_BUT           140.0   
12573    TXN_5167298     CUST_04      Furniture    Item_7_FUR           140.0   
12574    TXN_2407494     CUST_23           Food   Item_9_FOOD           170.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0         100.0       1850.0  Digital Wallet    Online       08/04/2024   
1          90.0       2610.0  Digital Wallet    Online       23/07/2023   
2          20.0        430.0     Credit Card    Online       05/10/2022   
3          90.0       2475.0     Credit Card    Online       07/05/2022   
4          70.0        875.0  Digital Wallet    Online       02/10/2022   
...         ...          ...             ...       ...              ...   
12570      40.0       1520.0     Credit Card  In-store       03/09/2023   
12571      90.0        585.0            Cash    Online       12/08/2022   
12572     100.0       1400.0            Cash    Online       24/08/2024   
12573      60.0        840.0            Cash    Online       30/12/2023   
12574      30.0        510.0            Cash    Online       06/08/2022   

       Discount Applied  
0                     1  
1                     1  
2                     0  
3                     0  
4                     0  
...                 ...  
12570                 0  
12571                 0  
12572                 0  
12573                 1  
12574                 0  

[11362 rows x 11 columns]

convierte toda la columna en una variable binaria, donde `1` **significa que se aplicó el descuento** y `0` que **no se aplicó**, dejando la columna lista para análisis cuantitativo.

In [52]:
# Revisión post-tratamiento
df.isna().sum()


Transaction ID      0
Customer ID         0
Category            0
Item                0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
dtype: int64

In [53]:
df.shape

(11362, 11)

### **Etapa 4 - Estandarización**

El prpósito de esta etapa es el de unificar formatos y dominios de los datos para que sean consistentes y comparables. La estandarización evita que una misma entidad aparezca registrada de distintas formas, lo que suele generar duplicados o errores en el análisis.

**Aspectos clave en este dataset**

* **Fechas (`Transaction Date`)**: ya están parseadas a `datetime`, pero conviene normalizarlas a un mismo estándar (ISO `YYYY-MM-DD`) y verificar que no haya valores atípicos fuera del rango esperado.  
* **Texto en categorías y productos (`Category`, `Item`)**: limpiar tildes, espacios extra y variaciones de mayúsculas/minúsculas. Por ejemplo, evitar que “Café”, “Cafe” y “cafe” se traten como categorías diferentes.  
* **Variables booleanas (`Discount Applied`)**: ya se estandarizó a 0/1, lo que permite un uso uniforme en cálculos y estadísticas.  
* **Unidades y coherencia**: confirmar que `Total Spent = Price Per Unit * Quantity`. Si no coincide, debe corregirse o marcarse.  


In [54]:
# Normalización de texto

# Crear copia explícita para evitar SettingWithCopyWarning
df = df.copy()

# Normalización de texto
import unicodedata, re  # para normalización de texto

def canon_texto(s):
    if pd.isna(s):
        return s
    s = unicodedata.normalize("NFKD", s).encode("ascii", "ignore").decode("ascii")  # quitar tildes
    return re.sub(r"\s+", " ", s.strip()).title()  # limpiar espacios y capitalizar

# aplicar normalización
df["Category"] = df["Category"].map(canon_texto)  # normalizar categorías
df["Item"] = df["Item"].map(canon_texto)          # normalizar ítems
df.head(10)

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0     TXN_6867343     CUST_09     Patisserie   Item_10_Pat           185.0   
1     TXN_3731986     CUST_22  Milk Products  Item_17_Milk           290.0   
2     TXN_9303719     CUST_02       Butchers   Item_12_But           215.0   
3     TXN_9458126     CUST_06      Beverages   Item_16_Bev           275.0   
4     TXN_4575373     CUST_05           Food   Item_6_Food           125.0   
6     TXN_3652209     CUST_07           Food   Item_1_Food            50.0   
8     TXN_9728486     CUST_23      Furniture   Item_16_Fur           275.0   
9     TXN_2722661     CUST_25       Butchers   Item_22_But           365.0   
10    TXN_8776416     CUST_22       Butchers    Item_3_But            80.0   
12    TXN_5874772     CUST_23           Food   Item_2_Food            65.0   

    Quantity  Total Spent  Payment Method  Location Transaction Date  \
0      100.0       1850.0  Digital Wallet    Online       08/04/2024   
1       90.0       2610.0  Digital Wallet    Online       23/07/2023   
2       20.0        430.0     Credit Card    Online       05/10/2022   
3       90.0       2475.0     Credit Card    Online       07/05/2022   
4       70.0        875.0  Digital Wallet    Online       02/10/2022   
6       80.0        400.0     Credit Card  In-store       10/06/2023   
8       10.0        275.0     Credit Card  In-store       26/04/2023   
9       30.0       1095.0            Cash    Online       14/03/2024   
10      90.0        720.0            Cash  In-store       14/12/2024   
12      70.0        455.0            Cash    Online       09/09/2023   

    Discount Applied  
0                  1  
1                  1  
2                  0  
3                  0  
4                  0  
6                  1  
8                  0  
9                  0  
10                 1  
12                 1

In [55]:
# Formato uniforme de fechas

df["Transaction Date"] = pd.to_datetime(
    df["Transaction Date"],
    errors="coerce",    # convierte lo inválido en NaT
    dayfirst=True       # pon False si las fechas están en formato mes/día/año
)

# Verifica
print(df["Transaction Date"].dtype)
print(df["Transaction Date"].head())


datetime64[ns]
0   2024-04-08
1   2023-07-23
2   2022-10-05
3   2022-05-07
4   2022-10-02
Name: Transaction Date, dtype: datetime64[ns]


In [56]:
# Verificación de coherencia monetaria

inconsistencias = df[df["Total Spent"] != df["Price Per Unit"] * df["Quantity"]]
print("Inconsistencias detectadas:", len(inconsistencias))


Inconsistencias detectadas: 11362


El resultado muestra un problema serio: en **11 362 de las 12 575 transacciones**, el campo `Total Spent` no coincide con `Price Per Unit * Quantity`. Esto indica que el valor está mal calculado o cargado en el dataset original. Para el análisis no se debe usar ese campo tal como viene, sino **reconstruirlo directamente a partir de precio unitario y cantidad**, garantizando así la coherencia de los indicadores de ingresos.


In [57]:
# Recalcular Total Spent como precio unitario * cantidad
df["Total Spent"] = df["Price Per Unit"] * df["Quantity"]

# Verificar de nuevo inconsistencias
inconsistencias = df[df["Total Spent"] != df["Price Per Unit"] * df["Quantity"]]
print("Inconsistencias detectadas después de la corrección:", len(inconsistencias))
df.head(10)

Inconsistencias detectadas después de la corrección: 0


Transaction ID Customer ID       Category          Item  Price Per Unit  \
0     TXN_6867343     CUST_09     Patisserie   Item_10_Pat           185.0   
1     TXN_3731986     CUST_22  Milk Products  Item_17_Milk           290.0   
2     TXN_9303719     CUST_02       Butchers   Item_12_But           215.0   
3     TXN_9458126     CUST_06      Beverages   Item_16_Bev           275.0   
4     TXN_4575373     CUST_05           Food   Item_6_Food           125.0   
6     TXN_3652209     CUST_07           Food   Item_1_Food            50.0   
8     TXN_9728486     CUST_23      Furniture   Item_16_Fur           275.0   
9     TXN_2722661     CUST_25       Butchers   Item_22_But           365.0   
10    TXN_8776416     CUST_22       Butchers    Item_3_But            80.0   
12    TXN_5874772     CUST_23           Food   Item_2_Food            65.0   

    Quantity  Total Spent  Payment Method  Location Transaction Date  \
0      100.0      18500.0  Digital Wallet    Online       2024-04-08   
1       90.0      26100.0  Digital Wallet    Online       2023-07-23   
2       20.0       4300.0     Credit Card    Online       2022-10-05   
3       90.0      24750.0     Credit Card    Online       2022-05-07   
4       70.0       8750.0  Digital Wallet    Online       2022-10-02   
6       80.0       4000.0     Credit Card  In-store       2023-06-10   
8       10.0       2750.0     Credit Card  In-store       2023-04-26   
9       30.0      10950.0            Cash    Online       2024-03-14   
10      90.0       7200.0            Cash  In-store       2024-12-14   
12      70.0       4550.0            Cash    Online       2023-09-09   

    Discount Applied  
0                  1  
1                  1  
2                  0  
3                  0  
4                  0  
6                  1  
8                  0  
9                  0  
10                 1  
12                 1

Este resultado confirma que al **recalcular `Total Spent` como `Price Per Unit * Quantity`**, todas las inconsistencias desaparecieron. Ahora la relación monetaria es coherente en todo el dataset, lo que asegura que los indicadores que dependan de montos estarán basados en datos consistentes y confiables.


### **Etapa 5 - Transformaciones**

En esta fase no se trata de limpiar, sino de **enriquecer** el dataset creando nuevas variables que agregan valor al análisis. El objetivo es homogeneizar criterios de negocio (ej. ingresos netos, métricas unitarias) y preparar datos que faciliten reportes y modelado posterior.

#### **Aspectos clave en este dataset**

* **Ingreso neto**: considerar el descuento aplicado (0/1) sobre el gasto total.  
* **Precio unitario neto**: dividir el ingreso neto entre la cantidad, para obtener el precio efectivo pagado por unidad.  
* **Variables temporales**: extraer de la fecha atributos útiles como mes, trimestre, año o día de la semana.  
* **Margen (si hubiera costos)**: no lo tenemos en este dataset, pero se podría calcular en un escenario real.  

A continuación vamos a realizar algunas acciones que, en este contexto, se pretende enriquezcan el dataset para mejorar el análisis:



In [58]:
# Crear monto neto según descuento aplicado:
# Asumimos que si Discount Applied = 1, se descuenta un **10%** del gasto total

df["Net Spent"] = df["Total Spent"] * (1 - 0.10 * df["Discount Applied"])
df.head(10)

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0     TXN_6867343     CUST_09     Patisserie   Item_10_Pat           185.0   
1     TXN_3731986     CUST_22  Milk Products  Item_17_Milk           290.0   
2     TXN_9303719     CUST_02       Butchers   Item_12_But           215.0   
3     TXN_9458126     CUST_06      Beverages   Item_16_Bev           275.0   
4     TXN_4575373     CUST_05           Food   Item_6_Food           125.0   
6     TXN_3652209     CUST_07           Food   Item_1_Food            50.0   
8     TXN_9728486     CUST_23      Furniture   Item_16_Fur           275.0   
9     TXN_2722661     CUST_25       Butchers   Item_22_But           365.0   
10    TXN_8776416     CUST_22       Butchers    Item_3_But            80.0   
12    TXN_5874772     CUST_23           Food   Item_2_Food            65.0   

    Quantity  Total Spent  Payment Method  Location Transaction Date  \
0      100.0      18500.0  Digital Wallet    Online       2024-04-08   
1       90.0      26100.0  Digital Wallet    Online       2023-07-23   
2       20.0       4300.0     Credit Card    Online       2022-10-05   
3       90.0      24750.0     Credit Card    Online       2022-05-07   
4       70.0       8750.0  Digital Wallet    Online       2022-10-02   
6       80.0       4000.0     Credit Card  In-store       2023-06-10   
8       10.0       2750.0     Credit Card  In-store       2023-04-26   
9       30.0      10950.0            Cash    Online       2024-03-14   
10      90.0       7200.0            Cash  In-store       2024-12-14   
12      70.0       4550.0            Cash    Online       2023-09-09   

    Discount Applied  Net Spent  
0                  1    16650.0  
1                  1    23490.0  
2                  0     4300.0  
3                  0    24750.0  
4                  0     8750.0  
6                  1     3600.0  
8                  0     2750.0  
9                  0    10950.0  
10                 1     6480.0  
12                 1     4095.0

In [59]:
# Calcular precio unitario neto

df["Net Price Per Unit"] = df["Net Spent"] / df["Quantity"]
df.head(10)

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0     TXN_6867343     CUST_09     Patisserie   Item_10_Pat           185.0   
1     TXN_3731986     CUST_22  Milk Products  Item_17_Milk           290.0   
2     TXN_9303719     CUST_02       Butchers   Item_12_But           215.0   
3     TXN_9458126     CUST_06      Beverages   Item_16_Bev           275.0   
4     TXN_4575373     CUST_05           Food   Item_6_Food           125.0   
6     TXN_3652209     CUST_07           Food   Item_1_Food            50.0   
8     TXN_9728486     CUST_23      Furniture   Item_16_Fur           275.0   
9     TXN_2722661     CUST_25       Butchers   Item_22_But           365.0   
10    TXN_8776416     CUST_22       Butchers    Item_3_But            80.0   
12    TXN_5874772     CUST_23           Food   Item_2_Food            65.0   

    Quantity  Total Spent  Payment Method  Location Transaction Date  \
0      100.0      18500.0  Digital Wallet    Online       2024-04-08   
1       90.0      26100.0  Digital Wallet    Online       2023-07-23   
2       20.0       4300.0     Credit Card    Online       2022-10-05   
3       90.0      24750.0     Credit Card    Online       2022-05-07   
4       70.0       8750.0  Digital Wallet    Online       2022-10-02   
6       80.0       4000.0     Credit Card  In-store       2023-06-10   
8       10.0       2750.0     Credit Card  In-store       2023-04-26   
9       30.0      10950.0            Cash    Online       2024-03-14   
10      90.0       7200.0            Cash  In-store       2024-12-14   
12      70.0       4550.0            Cash    Online       2023-09-09   

    Discount Applied  Net Spent  Net Price Per Unit  
0                  1    16650.0               166.5  
1                  1    23490.0               261.0  
2                  0     4300.0               215.0  
3                  0    24750.0               275.0  
4                  0     8750.0               125.0  
6                  1     3600.0                45.0  
8                  0     2750.0               275.0  
9                  0    10950.0               365.0  
10                 1     6480.0                72.0  
12                 1     4095.0                58.5

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11362 entries, 0 to 12574
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Transaction ID      11362 non-null  string        
 1   Customer ID         11362 non-null  string        
 2   Category            11362 non-null  object        
 3   Item                11362 non-null  object        
 4   Price Per Unit      11362 non-null  float64       
 5   Quantity            11362 non-null  float64       
 6   Total Spent         11362 non-null  float64       
 7   Payment Method      11362 non-null  string        
 8   Location            11362 non-null  string        
 9   Transaction Date    11362 non-null  datetime64[ns]
 10  Discount Applied    11362 non-null  int64         
 11  Net Spent           11362 non-null  float64       
 12  Net Price Per Unit  11362 non-null  float64       
dtypes: datetime64[ns](1), float64(5), int64(1), object(

In [61]:
# Extraer variables temporales

df["Year"] = df["Transaction Date"].dt.year
df["Month"] = df["Transaction Date"].dt.month
df["Weekday"] = df["Transaction Date"].dt.day_name()
df.head(10)

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0     TXN_6867343     CUST_09     Patisserie   Item_10_Pat           185.0   
1     TXN_3731986     CUST_22  Milk Products  Item_17_Milk           290.0   
2     TXN_9303719     CUST_02       Butchers   Item_12_But           215.0   
3     TXN_9458126     CUST_06      Beverages   Item_16_Bev           275.0   
4     TXN_4575373     CUST_05           Food   Item_6_Food           125.0   
6     TXN_3652209     CUST_07           Food   Item_1_Food            50.0   
8     TXN_9728486     CUST_23      Furniture   Item_16_Fur           275.0   
9     TXN_2722661     CUST_25       Butchers   Item_22_But           365.0   
10    TXN_8776416     CUST_22       Butchers    Item_3_But            80.0   
12    TXN_5874772     CUST_23           Food   Item_2_Food            65.0   

    Quantity  Total Spent  Payment Method  Location Transaction Date  \
0      100.0      18500.0  Digital Wallet    Online       2024-04-08   
1       90.0      26100.0  Digital Wallet    Online       2023-07-23   
2       20.0       4300.0     Credit Card    Online       2022-10-05   
3       90.0      24750.0     Credit Card    Online       2022-05-07   
4       70.0       8750.0  Digital Wallet    Online       2022-10-02   
6       80.0       4000.0     Credit Card  In-store       2023-06-10   
8       10.0       2750.0     Credit Card  In-store       2023-04-26   
9       30.0      10950.0            Cash    Online       2024-03-14   
10      90.0       7200.0            Cash  In-store       2024-12-14   
12      70.0       4550.0            Cash    Online       2023-09-09   

    Discount Applied  Net Spent  Net Price Per Unit  Year  Month    Weekday  
0                  1    16650.0               166.5  2024      4     Monday  
1                  1    23490.0               261.0  2023      7     Sunday  
2                  0     4300.0               215.0  2022     10  Wednesday  
3                  0    24750.0               275.0  2022      5   Saturday  
4                  0     8750.0               125.0  2022     10     Sunday  
6                  1     3600.0                45.0  2023      6   Saturday  
8                  0     2750.0               275.0  2023      4  Wednesday  
9                  0    10950.0               365.0  2024      3   Thursday  
10                 1     6480.0                72.0  2024     12   Saturday  
12                 1     4095.0                58.5  2023      9   Saturday

In [62]:
df["Net Spent"].describe()

count    11362.000000
mean     12526.935839
std       9170.217437
min        450.000000
25%       4950.000000
50%      10260.000000
75%      18270.000000
max      41000.000000
Name: Net Spent, dtype: float64

In [63]:
# Comparar estadísticas entre Total Spent y Net Spent
stats_comparacion = pd.DataFrame({
    "Total Spent": df["Total Spent"].describe(),
    "Net Spent": df["Net Spent"].describe()
})

print(stats_comparacion)


        Total Spent     Net Spent
count  11362.000000  11362.000000
mean   12964.253653  12526.935839
std     9466.236817   9170.217437
min      500.000000    450.000000
25%     5200.000000   4950.000000
50%    10850.000000  10260.000000
75%    19200.000000  18270.000000
max    41000.000000  41000.000000


### **Etapa 6 — De-duplicación**

El propósito de esta etapa es la de eliminar registros repetidos que distorsionan los indicadores de negocio. La duplicación puede ser **exacta** (todas las columnas iguales) o **parcial** (la misma transacción registrada dos veces con alguna variación mínima, como en el texto del ítem).

#### **Aspectos clave en este dataset**

* La clave de negocio es **`Transaction ID`**, que ya verificamos como única, por lo que no deberían existir duplicados a ese nivel.  
* Sin embargo, podrían existir duplicados **exactos** en todas las columnas (copias de la misma fila) o **duplicados “suaves”** (mismo cliente, ítem y fecha, pero distinto ID).  
* La limpieza busca identificar y cuantificar estos casos, para decidir si se eliminan o se consolidan.  


In [64]:
# Revisar duplicados exactos

dup_exactos = df.duplicated().sum()
print("Duplicados exactos:", dup_exactos)


Duplicados exactos: 0


No hay coincidencias en los registros, es decir, datos duplicados.  Si los hubiera, realizaríamos el siguiente paso:

In [65]:
# Eliminar duplicados exactos (si los hubiera)
df = df.drop_duplicates()


Podemos revisar también la aparición de duplicados parciales:

In [66]:
# Revisar duplicados parciales (ejemplo: mismo cliente + fecha + ítem)

dup_parciales = df.duplicated(subset=["Customer ID", "Transaction Date", "Item"]).sum()
print("Duplicados parciales (cliente-fecha-ítem):", dup_parciales)


Duplicados parciales (cliente-fecha-ítem): 11


De lo anterior: 

* **Duplicados exactos = 0** → no hay filas idénticas en todas las columnas, lo cual es positivo.  
* **Duplicados parciales = 11** → existen once casos en los que un mismo cliente, en la misma fecha y con el mismo ítem, aparece más de una vez.  

Esto último puede deberse a:  

* **Errores de captura** (la misma compra ingresada dos veces).  
* **Fragmentación de registros** (ej. el sistema divide en varias líneas una compra grande).  
* **Casos válidos de negocio** (ej. dos transacciones diferentes el mismo día con el mismo producto).  

Antes de decidir si eliminarlos, conviene **inspeccionarlos directamente** para ver de qué se trata.


#### **Etapa 6.1 — Inspección de duplicados parciales**


- **Definir la clave y marcar candidatos:** Buscamos repeticiones con la misma persona, fecha e ítem.


In [67]:
keys = ["Customer ID", "Transaction Date", "Item"]
dup_mask = df.duplicated(subset=keys, keep=False)
candidatos = df.loc[dup_mask].copy()

print("Filas candidatas:", len(candidatos))


Filas candidatas: 22


La inspección arrojó **22 filas candidatas a duplicados parciales**, es decir, transacciones con el mismo cliente, fecha e ítem registradas más de una vez. Esto no implica que sean copias idénticas, sino posibles divisiones de una misma compra o errores de captura que requieren revisión detallada.

- **Contar grupos y su tamaño** ¿Cuántos grupos hay y cuántas filas tiene cada uno?

In [68]:
grupos = (
    candidatos.groupby(keys)
    .size()
    .rename("n_rows")
    .reset_index()
    .sort_values("n_rows", ascending=False)
)

print("Grupos con posibles duplicados:", (grupos["n_rows"] > 1).sum())
grupos.head(10)


Grupos con posibles duplicados: 11


Customer ID Transaction Date          Item  n_rows
0     CUST_02       2024-01-12   Item_12_But       2
1     CUST_07       2022-06-30   Item_12_Cea       2
2     CUST_14       2023-03-06  Item_16_Milk       2
3     CUST_14       2023-04-05  Item_16_Milk       2
4     CUST_16       2024-05-23    Item_8_Fur       2
5     CUST_18       2022-11-13    Item_6_Ehe       2
6     CUST_20       2024-03-23   Item_16_But       2
7     CUST_21       2023-02-15    Item_1_Fur       2
8     CUST_22       2024-05-17  Item_13_Milk       2
9     CUST_22       2024-10-12   Item_12_Bev       2

El listado muestra **11 grupos de posibles duplicados**, cada uno con dos registros que comparten el mismo cliente, fecha e ítem. Esto sugiere que ciertas compras pudieron haberse fragmentado en más de una fila o registrarse por error. El siguiente paso es revisar el detalle de cada grupo (cantidades, precios y totales) para decidir si deben consolidarse o mantenerse como transacciones válidas.

- **Resumen por grupo (señales para decidir):** Precio unitario constante dentro del grupo suele indicar “partición” de una misma compra.


In [69]:
resumen = (
    candidatos.groupby(keys)
    .agg(
        n_rows      = ("Transaction ID", "count"),
        qty_sum     = ("Quantity", "sum"),
        total_sum   = ("Total Spent", "sum"),
        pu_nunique  = ("Price Per Unit", "nunique"),
        pu_min      = ("Price Per Unit", "min"),
        pu_max      = ("Price Per Unit", "max"),
    )
    .assign(pu_const=lambda d: d["pu_nunique"].eq(1))
    .sort_values(["n_rows","pu_const","qty_sum"], ascending=[False, False, False])
)

resumen.head(10)


n_rows  qty_sum  total_sum  \
Customer ID Transaction Date Item                                       
CUST_14     2023-04-05       Item_16_Milk       2    160.0    44000.0   
CUST_22     2024-10-12       Item_12_Bev        2    150.0    32250.0   
CUST_25     2022-06-23       Item_15_Cea        2    150.0    39000.0   
CUST_18     2022-11-13       Item_6_Ehe         2    130.0    16250.0   
CUST_20     2024-03-23       Item_16_But        2    130.0    35750.0   
CUST_21     2023-02-15       Item_1_Fur         2    120.0     6000.0   
CUST_22     2024-05-17       Item_13_Milk       2    100.0    23000.0   
CUST_07     2022-06-30       Item_12_Cea        2     90.0    19350.0   
CUST_02     2024-01-12       Item_12_But        2     60.0    12900.0   
CUST_14     2023-03-06       Item_16_Milk       2     60.0    16500.0   

                                           pu_nunique  pu_min  pu_max  \
Customer ID Transaction Date Item                                       
CUST_14     2023-04-05       Item_16_Milk           1   275.0   275.0   
CUST_22     2024-10-12       Item_12_Bev            1   215.0   215.0   
CUST_25     2022-06-23       Item_15_Cea            1   260.0   260.0   
CUST_18     2022-11-13       Item_6_Ehe             1   125.0   125.0   
CUST_20     2024-03-23       Item_16_But            1   275.0   275.0   
CUST_21     2023-02-15       Item_1_Fur             1    50.0    50.0   
CUST_22     2024-05-17       Item_13_Milk           1   230.0   230.0   
CUST_07     2022-06-30       Item_12_Cea            1   215.0   215.0   
CUST_02     2024-01-12       Item_12_But            1   215.0   215.0   
CUST_14     2023-03-06       Item_16_Milk           1   275.0   275.0   

                                           pu_const  
Customer ID Transaction Date Item                    
CUST_14     2023-04-05       Item_16_Milk      True  
CUST_22     2024-10-12       Item_12_Bev       True  
CUST_25     2022-06-23       Item_15_Cea       True  
CUST_18     2022-11-13       Item_6_Ehe        True  
CUST_20     2024-03-23       Item_16_But       True  
CUST_21     2023-02-15       Item_1_Fur        True  
CUST_22     2024-05-17       Item_13_Milk      True  
CUST_07     2022-06-30       Item_12_Cea       True  
CUST_02     2024-01-12       Item_12_But       True  
CUST_14     2023-03-06       Item_16_Milk      True

El resumen muestra que en todos los grupos de duplicados el **precio unitario es constante (`pu_const=True`)**, y lo que varía es la cantidad y el total gastado. Esto es un indicio claro de que se trata de **fragmentaciones de la misma compra** más que de transacciones distintas. En estos casos, la acción recomendada es **consolidar** cada grupo sumando cantidades y recalculando el gasto total, para evitar sobreestimar las ventas.

- **Ver el detalle del primer grupo:** Revisa manualmente si parecen “partes” de la misma compra o transacciones distintas.


In [70]:
g1 = resumen.query("n_rows > 1").head(1).reset_index()[keys].iloc[0]
cust, fecha, item = g1["Customer ID"], g1["Transaction Date"], g1["Item"]

detalle_g1 = (
    candidatos.loc[
        (candidatos["Customer ID"]==cust) &
        (candidatos["Transaction Date"]==fecha) &
        (candidatos["Item"]==item),
        ["Transaction ID","Customer ID","Transaction Date","Item",
         "Quantity","Price Per Unit","Total Spent","Discount Applied","Net Spent"]
    ]
    .sort_values(["Price Per Unit","Transaction ID"])
)
detalle_g1


Transaction ID Customer ID Transaction Date          Item  Quantity  \
11544    TXN_3478292     CUST_14       2023-04-05  Item_16_Milk      90.0   
2228     TXN_5990311     CUST_14       2023-04-05  Item_16_Milk      70.0   

       Price Per Unit  Total Spent  Discount Applied  Net Spent  
11544           275.0      24750.0                 0    24750.0  
2228            275.0      19250.0                 1    17325.0

#### **Interpretando estos resultados**

**1. Lo que muestra el resumen por grupo**

En todos los grupos de duplicados parciales, el indicador **`pu_const=True`**, lo que significa que el **precio unitario se mantiene constante** dentro de cada grupo. Esa es una señal de que los registros podrían corresponder a la **misma compra fragmentada en varias filas**, y no a transacciones distintas con precios cambiantes.

**2. Lo que muestra el detalle del primer grupo**

Al revisar el detalle de `CUST_14` el 2023-04-05 con `Item_16_Milk`, se ven dos registros con el mismo precio unitario, pero distinta **cantidad** y, sobre todo, diferencia en el campo **`Discount Applied`** (uno con descuento, otro sin él). Esto indica que aunque comparten precio, **no son idénticas en sus condiciones de venta**.

**3. Interpretación**

* **Casos claros para consolidar:** cuando `pu_const=True` y la única diferencia entre filas es la cantidad → se pueden sumar cantidades y recalcular el total.
* **Casos ambiguos:** cuando además del precio aparecen diferencias en condiciones como el descuento aplicado → deben analizarse aparte, ya que no se puede consolidar directamente sin una decisión de negocio (ej. consolidar con regla de promediar el descuento o mantener ambos registros separados).


El resumen indica que todos los grupos tienen precios unitarios consistentes, pero el detalle revela que no siempre son consolidables de forma automática: algunos son compras partidas, otros son compras con condiciones distintas.

#### **Regla automática para duplicados parciales**

##### **1. Identificar los grupos**

Agrupar por `Customer ID`, `Transaction Date`, `Item` y revisar:

* **Precio unitario (`pu_const`)**: debe ser constante.
* **Estado del descuento**: debe ser igual en todas las filas del grupo.

##### **2. Decisión**

* **Si `pu_const=True` y todos los registros del grupo tienen el mismo valor en `Discount Applied`** → **Consolidar**:

  * Sumar cantidades.
  * Recalcular `Total Spent = Price Per Unit * Quantity`.
  * Recalcular `Net Spent` aplicando la regla de descuento.
* **Si `pu_const=True` pero el descuento varía dentro del grupo** → **Ambiguo**: mantener las filas como están, marcadas con una bandera para revisión manual.
* **Si `pu_const=False`** (no es tu caso actual) → se asume que son **transacciones distintas**, no se consolidan.

##### **3. Ejemplo de implementación**

In [71]:
# === Consolidación robusta de duplicados parciales ===
keys = ["Customer ID", "Transaction Date", "Item"]

# Asegura el set completo de columnas en la salida
all_cols = list(df.columns)
if "Consolidated" not in all_cols:
    all_cols += ["Consolidated"]

def consolidar_grupo(grupo: pd.DataFrame) -> pd.DataFrame:
    same_pu   = (grupo["Price Per Unit"].nunique() == 1)
    same_disc = (grupo["Discount Applied"].nunique() == 1)
    # Consolidar solo si hay más de 1 fila y condiciones homogéneas
    if len(grupo) > 1 and same_pu and same_disc:
        first = grupo.iloc[0]
        # Partimos de "todas" las columnas del primer registro
        fila = first.to_dict()

        # Recalcular métricas agregadas
        qty   = grupo["Quantity"].sum()
        pu    = first["Price Per Unit"]
        disc  = first["Discount Applied"]
        total = pu * qty
        net   = total * (1 - 0.10 * disc)

        # Actualizar solo lo que cambia
        fila.update({
            # si quieres preservar trazabilidad, concatena IDs de las filas consolidadas:
            "Transaction ID": ";".join(grupo["Transaction ID"].astype(str)),
            "Quantity": qty,
            "Total Spent": total,
            "Net Spent": net,
            "Consolidated": True,
        })
        # Si existe columna derivada, mantener consistencia
        if "Net Price Per Unit" in fila:
            fila["Net Price Per Unit"] = (net / qty) if qty else pd.NA

        # Devolver con todas las columnas alineadas
        return pd.DataFrame([fila]).reindex(columns=all_cols)

    # Si no se consolida, devolver el grupo tal cual + bandera
    out = grupo.copy()
    out["Consolidated"] = False
    return out.reindex(columns=all_cols)

# Aplicación (compatible con pandas < 2.1; sin include_groups)
resultados = (
    df.groupby(keys, group_keys=False)
      .apply(consolidar_grupo)
      .reset_index(drop=True)
)


C:\Users\USER\AppData\Local\Temp\ipykernel_22176\1434034746.py:49: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(consolidar_grupo)


In [72]:
print(resultados["Consolidated"].value_counts(dropna=False))
print("Duplicados por clave en el resultado:", 
      resultados.duplicated(subset=keys, keep=False).sum())
resultados.isna().sum().sort_values(ascending=False).head(10)
resultados.info()

Consolidated
False    11348
True         7
Name: count, dtype: int64
Duplicados por clave en el resultado: 8
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11355 entries, 0 to 11354
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Transaction ID      11355 non-null  object        
 1   Customer ID         11355 non-null  object        
 2   Category            11355 non-null  object        
 3   Item                11355 non-null  object        
 4   Price Per Unit      11355 non-null  float64       
 5   Quantity            11355 non-null  float64       
 6   Total Spent         11355 non-null  float64       
 7   Payment Method      11355 non-null  object        
 8   Location            11355 non-null  object        
 9   Transaction Date    11355 non-null  datetime64[ns]
 10  Discount Applied    11355 non-null  int64         
 11  Net Spent           11355 non-null  float64      

In [73]:
resultados

Transaction ID Customer ID                            Category  \
0        TXN_6950467     CUST_01  Computers And Electric Accessories   
1        TXN_9662767     CUST_01                           Beverages   
2        TXN_5044628     CUST_01                            Butchers   
3        TXN_9131959     CUST_01                       Milk Products   
4        TXN_9753172     CUST_01  Computers And Electric Accessories   
...              ...         ...                                 ...   
11350    TXN_8216096     CUST_25       Electric Household Essentials   
11351    TXN_6314774     CUST_25                       Milk Products   
11352    TXN_4454614     CUST_25                          Patisserie   
11353    TXN_2596812     CUST_25                          Patisserie   
11354    TXN_5907338     CUST_25                           Beverages   

               Item  Price Per Unit  Quantity  Total Spent  Payment Method  \
0        Item_6_Cea           125.0      60.0       7500.0  Digital Wallet   
1       Item_20_Bev           335.0      90.0      30150.0     Credit Card   
2       Item_25_But           410.0      90.0      36900.0  Digital Wallet   
3       Item_2_Milk            65.0      10.0        650.0     Credit Card   
4        Item_9_Cea           170.0      90.0      15300.0  Digital Wallet   
...             ...             ...       ...          ...             ...   
11350   Item_21_Ehe           350.0      10.0       3500.0            Cash   
11351  Item_23_Milk           380.0      80.0      30400.0     Credit Card   
11352   Item_10_Pat           185.0      10.0       1850.0     Credit Card   
11353    Item_2_Pat            65.0      80.0       5200.0            Cash   
11354   Item_23_Bev           380.0      90.0      34200.0            Cash   

       Location Transaction Date  Discount Applied  Net Spent  \
0      In-store       2022-01-01                 1     6750.0   
1      In-store       2022-01-02                 1    27135.0   
2      In-store       2022-01-07                 0    36900.0   
3        Online       2022-01-11                 0      650.0   
4      In-store       2022-01-12                 1    13770.0   
...         ...              ...               ...        ...   
11350  In-store       2025-01-06                 0     3500.0   
11351    Online       2025-01-08                 0    30400.0   
11352  In-store       2025-01-13                 1     1665.0   
11353    Online       2025-01-17                 0     5200.0   
11354  In-store       2025-01-18                 1    30780.0   

       Net Price Per Unit  Year  Month    Weekday  Consolidated  
0                   112.5  2022      1   Saturday         False  
1                   301.5  2022      1     Sunday         False  
2                   410.0  2022      1     Friday         False  
3                    65.0  2022      1    Tuesday         False  
4                   153.0  2022      1  Wednesday         False  
...                   ...   ...    ...        ...           ...  
11350               350.0  2025      1     Monday         False  
11351               380.0  2025      1  Wednesday         False  
11352               166.5  2025      1     Monday         False  
11353                65.0  2025      1     Friday         False  
11354               342.0  2025      1   Saturday         False  

[11355 rows x 17 columns]

### **Etapa 7 — Verificación y enriquecimiento**

Antes de exportar, garantizamos **calidad y coherencia**: unicidad de claves, dominios válidos, rangos razonables, y reglas de negocio (p. ej., `Total Spent = Price Per Unit * Quantity`). Luego **enriquecemos** con variables de calendario que facilitan el análisis (año, mes, trimestre, fin de semana, etc.).


#### **7.1 Seleccionar el dataset base:** Se define con cuál DataFrame trabajar: si existe un `final_df` (tras consolidar duplicados), se toma; de lo contrario, se usa `df`. Esto evita confusiones y garantiza consistencia en el flujo.

In [74]:
final_df = resultados.copy()

# elige automáticamente el que exista
base = final_df if "final_df" in globals() else df
print(base.shape)

(11355, 17)


#### **7.2 Validaciones clave (unicidad, dominios, coherencia):** Aquí se revisa la **integridad de los datos**:

* **Unicidad**: que la clave de transacción o combinación de campos no tenga duplicados.  
* **Dominios válidos**: verificar que cantidades, precios y totales no sean negativos, y que `Discount Applied` tenga solo 0 o 1.  
* **Coherencia monetaria**: confirmar que `Total Spent = Price Per Unit × Quantity`.  


In [75]:
# 7.2.1 Unicidad (elige la mejor clave disponible)
if "Transaction ID" in base.columns:
    dup_tx = int(base["Transaction ID"].duplicated().sum())
    print("Duplicados por Transaction ID:", dup_tx)
else:
    dup_combo = int(base.duplicated(subset=["Customer ID","Transaction Date","Item","Discount Applied"]).sum())
    print("Duplicados por clave compuesta:", dup_combo)

# 7.2.2 Dominios/rangos básicos
checks = {
    "qty_nonneg":       bool((base["Quantity"] >= 0).all()),
    "price_nonneg":     bool((base["Price Per Unit"] >= 0).all()),
    "total_nonneg":     bool((base["Total Spent"] >= 0).all()),
    "discount_binary":  bool(set(base["Discount Applied"].unique()).issubset({0,1}))
}
checks

# 7.2.3 Coherencia monetaria (ya la corregimos, debería dar 0)
bad_money = int((base["Total Spent"] != base["Price Per Unit"]*base["Quantity"]).sum())
print("Inconsistencias monetarias:", bad_money)

Duplicados por Transaction ID: 0
Inconsistencias monetarias: 0


#### **7.3 Señalización de outliers (IQR simple, no elimina):** Se señalan (no eliminan) valores extremos usando el rango intercuartílico (IQR). Esto permite identificar compras anómalas (cantidades o montos inusualmente altos/bajos) que podrían ser errores o casos excepcionales.

In [76]:
import numpy as np

def flag_iqr(s, k=1.5):
    q1, q3 = s.quantile(0.25), s.quantile(0.75)
    iqr = q3 - q1
    return (s < (q1 - k*iqr)) | (s > (q3 + k*iqr))

base["Outlier_Quantity"] = flag_iqr(base["Quantity"])
base["Outlier_Price"]    = flag_iqr(base["Price Per Unit"])
base["Outlier_Total"]    = flag_iqr(base["Total Spent"])

base[["Outlier_Quantity","Outlier_Price","Outlier_Total"]].sum()

Outlier_Quantity     0
Outlier_Price        0
Outlier_Total       56
dtype: int64

#### **7.4 Enriquecimiento de calendario:** Se derivan variables temporales útiles: año, mes, trimestre, nombre del día y si la transacción fue en fin de semana. Esto facilita análisis posteriores como estacionalidad o patrones de consumo.

In [77]:
# Recalcula si no existen todavía
if "Year" not in base.columns:
    base["Year"] = base["Transaction Date"].dt.year
if "Month" not in base.columns:
    base["Month"] = base["Transaction Date"].dt.month
if "Weekday" not in base.columns:
    base["Weekday"] = base["Transaction Date"].dt.day_name()

base["Quarter"]    = base["Transaction Date"].dt.to_period("Q").astype(str)
base["YearMonth"]  = base["Transaction Date"].dt.to_period("M").astype(str)
base["IsWeekend"]  = base["Transaction Date"].dt.weekday >= 5  # sáb(5), dom(6)

base[["Year","Month","Quarter","YearMonth","Weekday","IsWeekend"]].head(3)

Year  Month Quarter YearMonth   Weekday  IsWeekend
0  2022      1  2022Q1   2022-01  Saturday       True
1  2022      1  2022Q1   2022-01    Sunday       True
2  2022      1  2022Q1   2022-01    Friday      False

#### **7.5 Resumen de calidad “antes de exportar”:** Se construye un reporte sintético: número de filas, campos con más nulos, duplicados, cumplimiento de reglas básicas y conteo de outliers. Es el **chequeo final de control de calidad** antes de exportar los datos limpios y enriquecidos.


In [78]:
quality_report = {
    "rows": len(base),
    "nulls_top": base.isna().sum().sort_values(ascending=False).head(8).to_dict(),
    "dups_tx_or_key": dup_tx if "Transaction ID" in base.columns else dup_combo,
    "range_checks": checks,
    "outliers": {
        "qty": int(base["Outlier_Quantity"].sum()),
        "price": int(base["Outlier_Price"].sum()),
        "total": int(base["Outlier_Total"].sum())
    }
}
quality_report

{'rows': 11355,
 'nulls_top': {'Transaction ID': 0,
  'Customer ID': 0,
  'Category': 0,
  'Item': 0,
  'Price Per Unit': 0,
  'Quantity': 0,
  'Total Spent': 0,
  'Payment Method': 0},
 'dups_tx_or_key': 0,
 'range_checks': {'qty_nonneg': True,
  'price_nonneg': True,
  'total_nonneg': True,
  'discount_binary': True},
 'outliers': {'qty': 0, 'price': 0, 'total': 56}}

### **Etapa 8 — Exportación de datos**

El propósito de este último paso consiste en guardar el dataset limpio en un formato adecuado para que pueda ser reutilizado en análisis, modelos o tableros de BI. El objetivo es **asegurar reproducibilidad, eficiencia y compatibilidad** con las herramientas de consumo.

#### **Recomendaciones técnicas**

1. **Formato de salida**

   * **Parquet** → ideal para analítica y grandes volúmenes de datos: comprimido, rápido de leer y conserva tipos de datos.
   * **CSV** → útil si necesitas intercambiar archivos con usuarios menos técnicos o sistemas que no leen Parquet, aunque pierde algo de eficiencia.

2. **Estructura de carpetas y versionado**

   * Guardar en rutas organizadas, por ejemplo:

     * `out/fact_ventas/anio=2023/mes=04/data.parquet`
     * `out/dim_clientes/current/data.parquet`
   * Esto facilita la integración con modelos estrella y sistemas de orquestación (Airflow, dbt, etc.).

3. **Codificación y separador** (si usas CSV)

   * Codificación: `utf-8`
   * Separador: `;` o `,` según convención regional
   * En BI suele trabajarse con **punto decimal interno**, dejando la **coma solo para reportes de presentación**.


In [80]:
# Exportar a formato CSV (alternativo)
df.to_csv("datasets/ventas_limpias.csv", index=False, sep=";", encoding="utf-8")


Con esto, el pipeline queda cerrado: **desde la importación hasta la exportación**, pasando por integración, limpieza, estandarización, transformaciones y deduplicación.
